### ISSUE TITLE
Stack overflow error in swagger validate

### ISSUE THREAD

**Issue by lafriks** [go-swagger/go-swagger#1614]
When running `swagger validate ./public/swagger.v1.json` probably after #1608 was merged we are having error:
```
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow
```

For more details see: https://drone.gitea.io/go-gitea/gitea/1918/8

**Comment by fredbi** [comment]
ok please join the spec to your issue so I may have a look at it

**Comment by lafriks** [comment]
Spec can be found here: https://github.com/go-gitea/gitea/blob/master/public/swagger.v1.json

Also just verified that when go-swagger is built from commit 036304b99147bbf483e6bb48e36697ea56d9fed1 it does work and give result:
```
2018/07/10 16:52:01
The swagger spec at "./public/swagger.v1.json" is valid against swagger specification 2.0
2018/07/10 16:52:01
The swagger spec at "./public/swagger.v1.json" showed up some valid but possibly unwanted constructs.
2018/07/10 16:52:01 See warnings below:
2018/07/10 16:52:01 - WARNING: response "#/responses/parameterBodies" is not used anywhere
2018/07/10 16:52:01 - WARNING: response "#/responses/redirect" is not used anywhere
2018/07/10 16:52:01 - WARNING: definition "#/definitions/CreateOrgOption" is not used anywhere
```

**Comment by fredbi** [comment]
ok. It is probably this which caused the regression https://github.com/go-openapi/spec/pull/81

**Comment by fredbi** [comment]
I can see the specs expands well, with expected circular $refs

**Comment by fredbi** [comment]
I could locate where it hangs: on ExpandResponse() for a response with a circular ref.
We have overlooked this subcase.

**Comment by pytlesk4** [comment]
I was going to push a fix for expand response, I don't know if it will cover this issue though, I can try resolving this spec.

**Comment by fredbi** [comment]
thanks @pytlesk4  - this PR exhausted me last week end. If you need to discuss issues with that, you can reach me on the slack

**Comment by fredbi** [comment]
I just tested ExpandSpec with this and it works fine.
The ExpandSchema used in validate is okay too.
I could find a hang on ExpandResponse for this response definition: `/repo/{owner}/{repo}/labels/{id}`

**Comment by fredbi** [comment]
@lafriks testing the new patch in preparation
also not that your spec _is_ invalid, but this a point overlooked by the validation:
pointing a schema to a response object (as in ``"204": "{ "description": "user is a member", "schema": { "$ref": "#/responses/empty"}}}` is not supported

**Comment by gregmarr** [comment]
@fredbi  Looks valid to me:
https://swagger.io/specification/v2/#responsesObject

    responses:
      '200':
        description: a pet to be returned
        schema:
          $ref: '#/definitions/Pet'
      default:
        description: Unexpected error
        schema:
          $ref: '#/definitions/ErrorModel'

**Comment by fredbi** [comment]
yes that is ok

**Comment by fredbi** [comment]
problem is
```yaml
responses:
  blah:
    description: "blah"

...
...
    responses:
      200:
         description: "peek a boo"
         schema:
           $ref: '#/responses/blah'
```

**Comment by fredbi** [comment]
because the element in responses is unmarshalled as a Response object not a Schema object

**Comment by fredbi** [comment]
see #1602  - I made a first version which was more tolerant, and eventually we figured out it was better to consider it an error.

**Comment by fredbi** [comment]
There is another codegen issue with this spec when generating the URL builder, with a name conflict on GetComments for String. This one cannot be a regression, the URLBuilder hasn't change in months

**Comment by gregmarr** [comment]
This part?

    responses:
      401ErrorResponse:
        description: An unexpected 401 error occurred.
        schema:
          $ref: "#/definitions/Error"
      404ErrorResponse:
        description: An unexpected 404 error occurred.
        schema:
          $ref: "#/definitions/Error"
      500ErrorResponse:
        description: An unexpected server error occurred.
        schema:
          $ref: "#/definitions/Error"

**Comment by pytlesk4** [comment]
The issue with the provided spec is that the schema ref is pointing at a response object.

```
schema:
           $ref: '#/responses/blah'
```

**Comment by fredbi** [comment]
yes @gregmarr  this one. There are several ones like this.
I think the previous version of flatten was more blind with this but my version is more picky and expects to get schemas to be analyzed.

**Comment by fredbi** [comment]
I thought it was detected by validate but it looks it isn't

**Comment by fredbi** [comment]
that's because expansion works a little differently, I tries to unmarshal again, so if this json can be put into a schema it does it.

**Comment by fredbi** [comment]
So I added a modified version of the original gitea spec in our CI run to work around this point and the name conflicting issue in urlbuilder. We'll see those as separate issues.

**Comment by fredbi** [comment]
OK so I admit the new go-openapi/analysis/Flatten() is pickier than it should:
```yaml
responses:
  blah:
    description: "blah"
```
is a valid candidate for a schema after all, when:
```yaml
responses:
  booh:
    description: "booh"
    schema:
       type: string
```
is not

I might change the Flatten() to attempt to unmarshall again as a schema.
Any thought?

**Comment by lafriks** [comment]
`name: string` was actually typo in our spec (should have been `name: since`)

Even with fixed gitea swagger json we are still getting same stack overflow.
Any pointers on what we should change in response descriptions to get validation working? :)

**Comment by lafriks** [comment]
@fredbi yes, I already tried running with modified taken from go-swagger CI fixtures:
go-swagger built from commit 8126ee94e1e178bbfd5561ac4d502d08446f47aa

```
swagger validate gitea-fixed.json
```

Takes much more time to get to overflow and at the end it is using about 7GB of RAM

**Comment by fredbi** [comment]
I just ran it on Linux ubutun 14, go1.9.  Latest commit.
Takes 0.2s seconds to report the warnings about unused $ref.

**Comment by fredbi** [comment]
Built at this commit:
```
commit 8126ee94e1e178bbfd5561ac4d502d08446f47aa
Merge: 82475a1 4d6f851
Author: Ivan Porto Carrero <ivan@flanders.co.nz>
Date:   Tue Jul 10 11:57:44 2018 -0700

    Merge pull request #1615 from fredbi/fix-1614-circular-in-response
    
    Fixes #1614

commit 4d6f851c3aeddb0548bc8aeeb84c2aae60a8a27a
Author: fredbi <frederic.bidon@advisam.consulting>
Date:   Tue Jul 10 19:59:35 2018 +0200

    Fixes #1614
```

**Comment by fredbi** [comment]
@lafriks really can't get to reproduce this. Ran it in loop many times.No way.
Could you please tell us more about your build/version etc?
